cmd/entrypoint/runner.go

```diff
@@ -50,13 +76,66 @@ func (rr *realRunner) Run(ctx context.Context, args ...string) error {
 	if rr.signals == nil {
 		rr.signals = make(chan os.Signal, 1)
 	}
-	defer close(rr.signals)
+	defer rr.close()
 	signal.Notify(rr.signals)
 	defer signal.Reset()
 
 	cmd := exec.CommandContext(ctx, name, args...)
+	stopCh := make(chan struct{}, 1)
+	defer close(stopCh)
+
 	cmd.Stdout = os.Stdout
+	var stdoutFile *os.File
+	if rr.stdoutPath != "" {
+		var err error
+		var doneCh <-chan error
+		if stdoutFile, err = os.Create(rr.stdoutPath); err != nil {
+			return err
+		}
+		// We use os.Pipe in asyncWriter to copy stdout instead of cmd.StdoutPipe or providing an
+		// io.Writer directly because otherwise Go would wait for the underlying fd to be closed by the
+		// child process before returning from cmd.Wait even if the process is no longer running. This
+		// would cause a deadlock if the child spawns a long running descendant process before exiting.
+		if cmd.Stdout, doneCh, err = asyncWriter(io.MultiWriter(os.Stdout, stdoutFile), stopCh); err != nil {
+			return err
+		}
+		go func() {
+			if err := <-doneCh; err != nil {
+				log.Fatalf("Copying stdout: %v", err)
+			}
+			stdoutFile.Close()
+		}()
+	}
+
 	cmd.Stderr = os.Stderr
+	var stderrFile *os.File
+	if rr.stderrPath != "" {
+		var err error
+		var doneCh <-chan error
+		if rr.stderrPath == rr.stdoutPath {
+			fd, err := syscall.Dup(int(stdoutFile.Fd()))
```

Comment:
Why copy stdout file descriptor and create a new file `stderrFile`? Can we rely on the already initialized file instance `stdoutFile`? I prefer implementing this way to keep both stdout and stderr independent of each other but just wondering if it's possible to reuse.

Reply:
I believe it is so that the file descriptors can be closed independently, but still point to the same file. Will do some additional checking to confirm.